File: api/src/main/java/com/doerapispring/domain/UserService.java
package com.doerapispring.domain;

import org.springframework.stereotype.Service;

import java.util.Optional;

@Service
public class UserService {
    private final ObjectRepository<User, UserId> userRepository;
    private final TodoListFactory todoListFactory;
    private final OwnedObjectRepository<TodoList, UserId, ListId> todoListRepository;

    UserService(ObjectRepository<User, UserId> userRepository,
                TodoListFactory todoListFactory,
                OwnedObjectRepository<TodoList, UserId, ListId> todoListRepository) {
        this.userRepository = userRepository;
        this.todoListFactory = todoListFactory;
        this.todoListRepository = todoListRepository;
    }

    public User create(String identifier) throws UserAlreadyExistsException {
        UserId userId = new UserId(identifier);
        Optional<User> userOptional = userRepository.find(userId);
        if (userOptional.isPresent()) throw new UserAlreadyExistsException();
        TodoList todoList = todoListFactory.todoList(
            userId,
            todoListRepository.nextIdentifier(),
            "default");
        User user = new User(userId, todoList.getListId());
        userRepository.save(user);
        todoListRepository.save(todoList);
        return user;
    }

    public Optional<User> find(String identifier) {
        return userRepository.find(new UserId(identifier));
    }
}

File: api/src/main/scala/com/doerapispring/domain/UserService.scala
package com.doerapispring.domain

import java.time.Clock
import java.util.{Date, Optional}

import org.springframework.stereotype.Service

@Service
class UserService(val userRepository: ObjectRepository[User, UserId],
                  val todoListFactory: TodoListFactory,
                  val todoListRepository: OwnedObjectRepository[TodoList, UserId, ListId],
                  val todoListModelSnapshotRepository: OwnedObjectWriteRepository[Snapshot[DeprecatedTodoListModel], UserId, ListId],
                  val clock: Clock) {
  def create(identifier: String): User = {
    val userId = new UserId(identifier)
    val userOptional = userRepository.find(userId)
    if (userOptional.isPresent) throw new UserAlreadyExistsException
    val todoList = todoListFactory.todoList(
      userId,
      todoListRepository.nextIdentifier(),
      "default")
    val listId = todoList.getListId
    val user = new User(userId, listId)
    userRepository.save(user)
    todoListRepository.save(todoList)
    todoListModelSnapshotRepository.save(
      user.getUserId,
      listId,
      Snapshot(DeprecatedTodoListModel(listId, todoList.getName, List(), new Date(0L), 0), Date.from(clock.instant())))
    user
  }

  def find(identifier: String): Optional[User] = {
    userRepository.find(new UserId(identifier))
  }
}